zerver/views/message_fetch.py

```diff
@@ -72,9 +72,19 @@ def get_search_fields(
     content_matches: Iterable[Tuple[int, int]],
     topic_matches: Iterable[Tuple[int, int]],
 ) -> Dict[str, str]:
+    # Following replacement is necessary because SQL-style HTML escaping of
+    # apostrophe (used to obtain `topic_matches`) differs from the django utils
+    # or any other library use to escape the apostrophe ('). This leads
+    # to topic highlighting in search results to be offset by one character.
+    quote_sql_html_escape_val = "&#x27;"
+    quote_django_html_escape_val = "&#39;"
+    escaped_topic_name = escape_html(topic_name).replace(
+        quote_sql_html_escape_val, quote_django_html_escape_val
+    )
```

Comment:
Thanks for tracking this down. Doing string replacements after escaping can often be dangerous from a security standpoint, though I think it's probably correct in this case. 

Given the [escaping algorithm](https://docs.python.org/3/library/html.html) [used by Django](https://docs.djangoproject.com/en/4.2/ref/utils/#module-django.utils.html), I think this might be OK.

@andersk FYI, is there a better way to do this without reimplementing one of these? I don't see a parameter in the Python one for how to escape `'`s or anything like that.

Reply:
It’d be more robust to change the SQL query to return the HTML-escaped topic name, escaped with the same function as the one used for `topic_matches`, so that we don’t have to carefully make two different escaping methods consistent with each other.